parser/src/main/scala/org/polystat/odin/parser/Utils.scala

```diff
@@ -0,0 +1,51 @@
+package org.polystat.odin.parser
+
+import com.github.tarao.nonempty.collection.NonEmpty
+import higherkindness.droste.data.Fix
+import org.polystat.odin.core.ast.astparams.EOExprOnly
+import org.polystat.odin.core.ast._
+
+object Utils {
+
+  def createArrayFromNonEmpty(ne: Option[NonEmpty[EOBnd[EOExprOnly], Vector[EOBnd[EOExprOnly]]]]): EOExprOnly = {
+    Fix[EOExpr](EOArray(
+      ne match {
+        case Some(value) => value
+        case None => Vector()
+      }
+    ))
+  }
+
+  def createNonEmpty(objs: Seq[EOBnd[EOExprOnly]])
+  : NonEmpty[EOBnd[EOExprOnly], Vector[EOBnd[EOExprOnly]]] = {
+    NonEmpty.from(objs) match {
+      case Some(value) => value.toVector
+      case None => throw new Exception("1 or more arguments expected, got 0.")
```

Comment:
I suggest to resolve this by getting rid of NonEmpty altogether. The only place where it is used is in the definition of EOCopy:
```scala
sealed case class EOCopy[+A](
  trg: A,
  args: NonEmpty[EOBnd[A], Vector[EOBnd[A]]]
) extends EOApp[A]
```
However, the rule which is responsible for parsing the application arguments already guarantees that only 1 or more arguments get through:
```scala
i => new Parser(indent = i).`object`
        .repX(1, sep = ("\n" ~ (" " * i))./)
```
`repX(1, ...)` basically means that the object should be parsed 1 or more times or else the parser will fail. So, the parser already guarantees that application arguments would not be empty, so I see no point in enforcing it once more using `NonEmpty`.

Reply:
No, we must have non-empty in the `EOCopy` for args, because otherwise `EOCopy(???, Vector())` will be equivalent to the `EOExpr` that is passed as a first argument to `EOCopy`. I would to restrict AST from having such ambiguity.

If you can guarantee that there will be elements to put in nonempty list you don't need to throw exceptions anyway.

Reply:
should be resolved [here](https://github.com/polystat/odin/pull/2/commits/d9ca4220da6bb6c0de9006f99c0816e8f9b85392)